scripts/package.json

```diff
@@ -2,7 +2,7 @@
   "version": "0.0.510771",
   "license": "SEE LICENSE IN https://chromium.googlesource.com/chromium/src/+/master/LICENSE",
   "scripts": {
-    "build-protocol-dts": "ts-node protocol-dts-generator.ts",
+    "build-protocol-dts": "ts-node --compiler ttypescript protocol-dts-generator.ts",
```

Comment:
I looked into tttypescript which seems like a fork of typescript for inline transformers, and typescript-is is the transformer plugin.

I do like the simplicity of `assertType<X>` but it seems like a lot of magic and dependencies that may not be maintained in the future.

I'm deffo down for upgrading TS version. Sounds like a great idea.

Since this is a static import we can use the `import` syntax for json and typescript will validate the types for us at compile time.  When I was writing the initial code, `import` syntax for json wasn't available yet.


![Uploading image.png…]()


```json
{
  "hello": "world",
  "foo": true,
  "bar": 0,
  "array": null
}
```

```ts
import testJson from './test.json';

interface Test {
  hello: string;
  foo: boolean;
  br: number;
}

const test: Test = testJson;
```

^ typescript will fail compilation

Reply:
I wasn't aware of the JSON `import` syntax! That seems a lot cleaner. :)

I did end up getting errors after I tried implementing it though:
```
protocol-dts-generator.ts(7,7): error TS2322: Type '{ version: { major: string; minor: string; }; domains: ({ domain: string; description: string; deprecated: boolean; dependencies: string[]; types: { id: string; description: string; type: string; properties: ({ ...; } | ... 1 more ... | { ...; })[]; }[]; commands: { ...; }[]; events: { ...; }[]; experimental?: undef...' is not assignable to type 'IProtocol'.
  Types of property 'domains' are incompatible.
    Type '({ domain: string; description: string; deprecated: boolean; dependencies: string[]; types: { id: string; description: string; type: string; properties: ({ name: string; description: string; type: string; enum: string[]; optional?: undefined; } | { ...; } | { ...; })[]; }[]; commands: { ...; }[]; events: { ...; }[];...' is not assignable to type 'Domain[]'.
      Type '{ domain: string; description: string; deprecated: boolean; dependencies: string[]; types: { id: string; description: string; type: string; properties: ({ name: string; description: string; type: string; enum: string[]; optional?: undefined; } | { ...; } | { ...; })[]; }[]; commands: { ...; }[]; events: { ...; }[]; ...' is not assignable to type 'Domain'.
        Type '{ domain: string; description: string; deprecated: boolean; dependencies: string[]; types: { id: string; description: string; type: string; properties: ({ name: string; description: string; type: string; enum: string[]; optional?: undefined; } | { ...; } | { ...; })[]; }[]; commands: { ...; }[]; events: { ...; }[]; ...' is not assignable to type 'Domain'.
          Types of property 'types' are incompatible.
            Type '{ id: string; description: string; type: string; properties: ({ name: string; description: string; type: string; enum: string[]; optional?: undefined; } | { name: string; description: string; type: string; enum?: undefined; optional?: undefined; } | { ...; })[]; }[]' is not assignable to type 'DomainType[]'.
              Type '{ id: string; description: string; type: string; properties: ({ name: string; description: string; type: string; enum: string[]; optional?: undefined; } | { name: string; description: string; type: string; enum?: undefined; optional?: undefined; } | { ...; })[]; }' is not assignable to type 'DomainType'.
                Type '{ id: string; description: string; type: string; properties: ({ name: string; description: string; type: string; enum: string[]; optional?: undefined; } | { name: string; description: string; type: string; enum?: undefined; optional?: undefined; } | { ...; })[]; }' is not assignable to type '{ id: string; description?: string | undefined; } & ObjectType'.
                  Type '{ id: string; description: string; type: string; properties: ({ name: string; description: string; type: string; enum: string[]; optional?: undefined; } | { name: string; description: string; type: string; enum?: undefined; optional?: undefined; } | { ...; })[]; }' is not assignable to type 'ObjectType'.
                    Types of property 'type' are incompatible.
                      Type 'string' is not assignable to type '"object"'.
protocol-dts-generator.ts(8,7): error TS2322: Type '{ version: { major: string; minor: string; }; domains: ({ domain: string; experimental: boolean; dependencies: string[]; types: ({ id: string; description: string; type: string; enum?: undefined; properties?: undefined; } | ... 4 more ... | { ...; })[]; commands: ({ ...; } | ... 1 more ... | { ...; })[]; events?: un...' is not assignable to type 'IProtocol'.
  Types of property 'domains' are incompatible.
    Type '({ domain: string; experimental: boolean; dependencies: string[]; types: ({ id: string; description: string; type: string; enum?: undefined; properties?: undefined; } | { id: string; description: string; type: string; enum: string[]; properties?: undefined; } | { ...; } | { ...; } | { ...; } | { ...; })[]; commands:...' is not assignable to type 'Domain[]'.
      Type '{ domain: string; experimental: boolean; dependencies: string[]; types: ({ id: string; description: string; type: string; enum?: undefined; properties?: undefined; } | { id: string; description: string; type: string; enum: string[]; properties?: undefined; } | { ...; } | { ...; } | { ...; } | { ...; })[]; commands: ...' is not assignable to type 'Domain'.
        Type '{ domain: string; experimental: boolean; dependencies: string[]; types: ({ id: string; description: string; type: string; enum?: undefined; properties?: undefined; } | { id: string; description: string; type: string; enum: string[]; properties?: undefined; } | { ...; } | { ...; } | { ...; } | { ...; })[]; commands: ...' is not assignable to type 'Domain'.
          Types of property 'types' are incompatible.
            Type '({ id: string; description: string; type: string; enum?: undefined; properties?: undefined; } | { id: string; description: string; type: string; enum: string[]; properties?: undefined; } | { id: string; description: string; type: string; properties: ({ ...; } | ... 1 more ... | { ...; })[]; enum?: undefined; } | { ....' is not assignable to type 'DomainType[]'.
              Type '{ id: string; description: string; type: string; enum?: undefined; properties?: undefined; } | { id: string; description: string; type: string; enum: string[]; properties?: undefined; } | { id: string; description: string; type: string; properties: ({ ...; } | ... 1 more ... | { ...; })[]; enum?: undefined; } | { .....' is not assignable to type 'DomainType'.
                Type '{ id: string; description: string; type: string; enum?: undefined; properties?: undefined; }' is not assignable to type 'DomainType'.
                  Type '{ id: string; description: string; type: string; enum?: undefined; properties?: undefined; }' is not assignable to type '{ id: string; description?: string | undefined; } & ObjectType'.
                    Type '{ id: string; description: string; type: string; enum?: undefined; properties?: undefined; }' is not assignable to type 'ObjectType'.
                      Types of property 'type' are incompatible.
                        Type 'string' is not assignable to type '"object"'.
```

If I'm interpreting this right, then when importing the JSON it's using types that are too wide (e.g. inferred: `type: string`) for the literal types we're specifying (actual: `type: "object"` in `ObjectType`). It doesn't look like we have a good way to resolve that - see: https://github.com/microsoft/TypeScript/issues/26552.

What do you propose we do?